### ISSUE TITLE
[Regression] apollo-client 3.0.0-beta.44 does not rollback optimistic updates

### ISSUE THREAD

**Issue by darkbasic** [apollographql/apollo-client#6183]
This is the repro: https://github.com/darkbasic/gh6183

After cloning use `yarn && yarn start` to run it. There is a button next to each article: if you delete the first two of them everything will work out as expected, but as soon as you will delete the third one the server will return an hardcoded error after a 3 seconds delay and apollo-client won't rollback the optimistic update.
If you run `git revert HEAD && yarn install:clean && yarn start` you will downgrade to apollo-client 2.6 and you will find that everything will work as expected: after 3 seconds apollo-client will rollback the failed mutation.

**Versions**

```
  System:
    OS: Linux 5.6 Arch Linux
  Binaries:
    Node: 13.13.0 - ~/.nvm/versions/node/v13.13.0/bin/node
    Yarn: 1.22.4 - /usr/bin/yarn
    npm: 6.14.4 - ~/.nvm/versions/node/v13.13.0/bin/npm
  Browsers:
    Firefox: 75.0
  npmPackages:
    @apollo/client: ^3.0.0-beta.44 => 3.0.0-beta.44
```

**Comment by darkbasic** [comment]
I suspect that re-accessing the same query with another `useQuery` will provide the correct result, thus the bug is probably just active queries not being updated. Neverthless this is not expected behaviour and a regression compared to apollo-client 2.6.

**Comment by caspersmith** [comment]
https://github.com/apollographql/apollo-client/issues/5790

**Comment by 3nvi** [comment]
Yeah I think this is a duplicate of #5790

**Comment by darkbasic** [comment]
I gave this another try with @beta.48 (https://github.com/apollographql/apollo-client/pull/6221), but unfortunately it still doesn't roll it back.

Worth mentioning, I also noticed another bad behaviour on the Apollo side: if I use `cache.evict` to delete the match (instead of `writeQuery`) it doesn't wait for the mutation response before re-fetching. Instead, as soon as it gets the optimistic reponse it immediately refetches from the server, which may or may not hold the correct response at this point. If you feel like this is worth its own bug report let me know and I'll open another one.
